src/AspNet.Security.OAuth.Airtable/AirtableAuthenticationHandler.cs

```diff
@@ -0,0 +1,137 @@
+        {
+            tokenRequestParameters.Add(OAuthConstants.CodeVerifierKey, codeVerifier!);
+            context.Properties.Items.Remove(OAuthConstants.CodeVerifierKey);
+        }
+
+        using var requestMessage = new HttpRequestMessage(HttpMethod.Post, Options.TokenEndpoint);
+        requestMessage.Headers.Accept.Add(
+            new MediaTypeWithQualityHeaderValue(MediaTypeNames.Application.FormUrlEncoded));
+        requestMessage.Content = new FormUrlEncodedContent(tokenRequestParameters);
+        var credentials = Convert.ToBase64String(Encoding.ASCII.GetBytes(
+            $"{Options.ClientId}:{Options.ClientSecret}"));
+        requestMessage.Headers.Authorization = new AuthenticationHeaderValue("Basic", credentials);
+
+        requestMessage.Version = Backchannel.DefaultRequestVersion;
+
+        var response = await Backchannel.SendAsync(requestMessage, Context.RequestAborted);
+        var body = await response.Content.ReadAsStringAsync(Context.RequestAborted);
+
+        return response.IsSuccessStatusCode switch
+        {
+            true => OAuthTokenResponse.Success(JsonDocument.Parse(body)),
+            false => await ParseInvalidResponseAsync(response)
+        };
+    }
+
+    private async Task<OAuthTokenResponse> ParseInvalidResponseAsync(HttpResponseMessage response)
+    {
+        await Log.ExchangeCodeErrorAsync(Logger, response, Context.RequestAborted);
+        return OAuthTokenResponse.Failed(new Exception("An error occurred while retrieving an access token."));
+    }
+
+    private static partial class Log
+    {
+        internal static async Task UserProfileErrorAsync(ILogger logger, HttpResponseMessage response, CancellationToken cancellationToken)
+        {
+            UserProfileError(
+                logger,
+                response.StatusCode,
+                response.Headers.ToString(),
+                await response.Content.ReadAsStringAsync(cancellationToken));
+        }
+
+        internal static async Task ExchangeCodeErrorAsync(ILogger logger, HttpResponseMessage response, CancellationToken cancellationToken)
+        {
+            ExchangeCodeError(
+                logger,
+                response.StatusCode,
+                response.Headers.ToString(),
+                await response.Content.ReadAsStringAsync(cancellationToken));
+        }
+
+        [LoggerMessage(3, LogLevel.Error, "An error occurred while retrieving an access token: the remote server returned a {Status} response with the following payload: {Headers} {Body}.")]
+        private static partial void ExchangeCodeError(
+            ILogger logger,
+            HttpStatusCode status,
+            string headers,
+            string body);
+
+        [LoggerMessage(1, LogLevel.Error,
+            "An error occurred while retrieving the user profile: the remote server returned a {Status} response with the following payload: {Headers} {Body}.")]
```

Comment:
Copy-paste artifact probably, but the IDs are 3 and then 1. Would make more sense to be 1 and 2.